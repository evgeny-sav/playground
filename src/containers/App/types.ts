import { Action } from 'redux';
import { ActionTypes } from './constants';

export interface AppState {
  data: any;
}

export interface AppProps {}

export interface FetchActionPayload {
  data: {
    id: number;
  };
}

export interface FetchSuccessActionPayload {
  data: any;
}

export interface FetchSuccessAction extends Action {
  payload: FetchSuccessActionPayload;
}

export interface FetchErrorAction extends Action {
  type: ActionTypes.FetchError;
  payload: string;
}

export type AppActions = FetchErrorAction | FetchSuccessAction;
