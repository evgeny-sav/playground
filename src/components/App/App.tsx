import React, { FC } from 'react';
import logo from '../../assets/logo.svg';
import './App.module.scss';
import { AppProps } from './types';

const App: FC<AppProps> = () => {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
      </header>
    </div>
  );
};

export default App;
