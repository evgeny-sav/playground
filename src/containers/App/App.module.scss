.App {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  width: 100%;
  min-height: 100vh;

  font-size: calc(10px + 2vmin);
  color: white;
  text-align: center;

  background-color: #282c34;

  p {
    img {
      border-radius: 50%;
    }
  }

}
