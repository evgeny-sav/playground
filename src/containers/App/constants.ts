export enum ActionTypes {
  Fetch = 'APP/FETCH',
  FetchError = 'APP/FETCH_ERROR',
  FetchSuccess = 'APP/FETCH_SUCCESS',
}
