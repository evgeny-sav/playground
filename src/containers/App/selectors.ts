import { Store } from '../../store/types';

export const appSelector = (store: Store) => store.app;
export const loadingSelector = (store: Store) => store.loading.isLoading;
