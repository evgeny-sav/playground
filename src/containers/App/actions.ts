import { Dispatch } from 'redux';
import {
  FetchActionPayload,
  FetchErrorAction,
  FetchSuccessAction,
  FetchSuccessActionPayload,
} from './types';
import { ActionTypes } from './constants';
import { loading } from '../../components/Loading/actions';

export const fetchStarted = (payload: FetchActionPayload) => ({
  type: ActionTypes.Fetch,
  payload,
});

export const fetchError = (e: string): FetchErrorAction => ({
  type: ActionTypes.FetchError,
  payload: e,
});

export const fetchSuccess = (
  payload: FetchSuccessActionPayload,
): FetchSuccessAction => ({
  type: ActionTypes.FetchSuccess,
  payload,
});

export const fetchAll = (data: FetchActionPayload) => async (
  dispatch: Dispatch,
) => {
  dispatch(loading(true));
  dispatch(fetchStarted(data));
  try {
    const response = await fetch('https://randomuser.me/api/');
    const { results } = await response.json();
    console.log('RESULTS', results);
    dispatch(fetchSuccess({ data: results }));
    dispatch(loading(false));
  } catch (e) {
    dispatch(fetchError(e.message));
    dispatch(loading(false));
  }
};
