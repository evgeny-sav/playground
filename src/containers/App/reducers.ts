import { Reducer } from 'redux';
import { ActionTypes } from './constants';
import {
  AppActions,
  AppState,
  FetchErrorAction,
  FetchSuccessAction,
} from './types';

export const initialState: AppState = {
  data: {},
};

const actions = {
  [ActionTypes.Fetch]: (state: any) => state,
  [ActionTypes.FetchError]: (_: any, action: FetchErrorAction) => {
    throw Error(action.payload);
  },
  [ActionTypes.FetchSuccess]: (_: any, action: FetchSuccessAction) =>
    action.payload,
};

const AppReducer: Reducer<AppState, AppActions> = (
  state = initialState,
  action: AppActions,
) => (actions[action.type] ? actions[action.type](state, action) : state);

export default AppReducer;
