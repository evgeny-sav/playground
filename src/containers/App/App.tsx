import React, { FC, useEffect } from 'react';
import styles from './App.module.scss';
import { AppProps } from './types';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { fetchAll } from './actions';
import { appSelector, loadingSelector } from './selectors';
import Loading from '../../components/Loading';

const App: FC<AppProps> = () => {
  const dispatch = useDispatch();
  const { data } = useSelector(appSelector, shallowEqual);
  const isLoading = useSelector(loadingSelector, shallowEqual);

  useEffect(() => {
    dispatch(fetchAll({ data: { id: 1 } }));
  }, []);

  return isLoading ? (
    <Loading />
  ) : (
    <div className={styles.App}>
      {data.map((el: any) => (
        <div key={el.id.value}>
          <p>
            <img src={el.picture.large} alt={el.name.first} />
          </p>
          <p>{`${el.name.title}. ${el.name.first} ${el.name.last}`}</p>
          <p>{`${el.location.city}`}</p>
          <p>{`${el.email}`}</p>
          <p>{`${el.cell}`}</p>
        </div>
      ))}
    </div>
  );
};

export default App;
