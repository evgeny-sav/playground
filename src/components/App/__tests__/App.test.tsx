import React from 'react';
import { create } from 'react-test-renderer';
import App from '../App';

test('renders learn react link', () => {
  const component = create(<App />);
  expect(component.toJSON()).toMatchSnapshot();
});
