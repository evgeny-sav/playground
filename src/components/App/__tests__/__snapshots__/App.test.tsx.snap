// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
  </header>
</div>
`;
